src/controllers/controller.h

```diff
@@ -79,6 +79,24 @@ class Controller : public QObject {
     void stopLearning();
 
   protected:
+    template<typename SpecificMappingType>
+    std::shared_ptr<SpecificMappingType> downcastAndTakeOwnership(
+            std::shared_ptr<LegacyControllerMapping> pMapping) {
+        // Controller cannot take ownership if pMapping is referenced elsewhere because
+        // the controller polling thread needs exclusive accesses to the non-thread safe
+        // LegacyControllerMapping.
+        // Trying to cast a std::shared_ptr to a std::unique_ptr is not worth the trouble.
+        // The use_count is 2 at this point, one use by the caller and one use by this class.
+        VERIFY_OR_DEBUG_ASSERT(pMapping.use_count() == 2) {
```

Comment:
had to change this to 2 because C++

wtf is this language :upside_down_face:

Reply:
If you pass it with `std::move` it should (or might?) work as expected. This is what you would also need to do when using a std::unique_ptr.

Reply:
Explicitly enforce the *move* on the caller site by turning the parameter type into an rvalue reference, i.e. `std::shared_ptr\<..>&&`.

Reply:
All ugly workarounds for missing move semantics and the reason why std::unique_ptr would be a better choice for the parameter type.

Reply:
Casting that magic spell worked :mage_man:

Reply:
Maybe I could make a separate `uniqueClone` method for LegacyControllerMapping? Also ugly, but :shrug:

Reply:
Whatever, this works and CI passed. Let's move on.